### PR TITLE
feat: Sync & Misc Fixes

### PR DESCRIPTION
**Pull request by reveriadive**
Fixes a bug with duplication, and actor sync.

### CODE REVIEW

**Comment by Pospelove** [comment]
Amazing job! I can imagine that took hours to debug. hope skymp team will take a look soon

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 1]
```diff
@@ -254,90 +246,69 @@ bool WorldState::AttachEspmRecord(const espm::CombineBrowser& br,
...
+	return false;
+  }
+  
+  auto existing = LookupFormById(formId);
+
+  // The root cause of duplication is here.
+  if (!existing) {
+	if (t == "NPC_" && typeStrEval == "ACHR") {
+	auto npcData = reinterpret_cast<espm::NPC_*>(base.rec)->GetData(cache);
+	//logger->info("{:#x} {:#x}", formId, worldOrCell);
```
Comment (nic11) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:301]: Probably not needed anymore?

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp [hunk 2]
```diff
@@ -31,9 +31,19 @@ std::string MpObjectReference::CreatePropertyMessage(
...
+  auto base = loader.GetBrowser().LookupById(GetBaseId());
+  if (base.rec) {
+    baseRecordType = base.rec->GetType().ToString();
+  }
+
   return nlohmann::json{ { "idx", self->GetIdx() },
                          { "t", MsgType::UpdateProperty },
                          { "propName", name },
+                         { "refrId", self->GetFormId() },
+                         { "baseRecordType", baseRecordType },
```
Comment (nic11) [inline skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:46]: Why do we need them? I see no usages in the client code

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp [hunk 3]
```diff
@@ -979,11 +989,9 @@ void MpObjectReference::ProcessActivate(MpObjectReference& activationSource)
     auto refrRecord = espm::Convert<espm::REFR>(
       loader.GetBrowser().LookupById(GetFormId()).rec);
     auto teleport = refrRecord->GetData(compressedFieldsCache).teleport;
-    if (teleport) {
-      if (!IsOpen()) {
-        SetOpen(true);
-        RequestReloot();
-      }
+	// Worldspace sync fix here.
```
Comment (nic11) [inline skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:992]: There's no need to write that in the code itself. It's better to either leave comment on the PR (like this one)

**Review by nic11 — COMMENTED** [review]
Had a very brief look on organizational matters. Let's discuss it a bit on Discord DMs, since I don't quite understand what you did here.

Please also:
1. Run autoformatting. Either clang-format itself or formatting option somewhere in Visual Studio menus
2. Rename the PR to something like `fix: did this and this` or `fix: resolve a few actor sync issues` (or their brief description, if you can make it into 5-10 words)

P.S. This is not [refactoring](https://en.wikipedia.org/wiki/Code_refactoring):
> without changing its external behavior

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp (hunk 3 above)
Comment (reveriadive) [reply skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:992]: You're right, it was a reminder for myself. I will do that for further commits.

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp (hunk 2 above)
Comment (reveriadive) [reply skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:46]: This must have been fixes I pulled from another repository. I wanted to merge the fixes by @Pospelove as they weren't in the original repository. (Without those fixes the game stutters a lot.)

**Comment by reveriadive** [comment]
Right now the regeneration system needs a little more work. (I came to some issues with my implementation of it.) 
At some point the health gets stuck at 0 without dying. Some enemies were invulnerable too.

Until that I would like if you didn't merge the request yet.

File: skymp5-server/cpp/server_guest_lib/ActionListener.cpp [hunk 4]
```diff
@@ -381,7 +380,8 @@ void UseCraftRecipe(MpActor* me, espm::COBJ::Data recipeData,
     spdlog::debug("{}", s);
   }
   me->RemoveItems(entries);
-  me->AddItem(outputFormId, recipeData.outputCount);
+  me->AddItem(espm::GetMappedId(recipeData.outputObjectFormId, *mapping),
+              recipeData.outputCount);
```
[the remark this answers is not in the document]
Comment (Pospelove) [reply skymp5-server/cpp/server_guest_lib/ActionListener.cpp:384]: I think it outputs wrong id originally. but let's wait for author's response

**Comment by reveriadive** [comment]
image
This should be good to go now. I'll make a new pull request for further changes to code.

**Comment by reveriadive** [comment]
The OnHit code throws an error at the moment when attacking npcs. I've narrowed the cause down to setting the target actor, GetFormAt() on line 680, but have no idea how to fix the invalid form on the server-side. (The client reports the actors to be the same id through console.)

**Comment by reveriadive** [comment]
image
The actor attacked is below.
ScreenShot9

**Comment by reveriadive** [comment]
Okay, so as it turns out that's because I was using the console to teleport. Guess the actor cache doesn't update.

**Comment by reveriadive** [comment]
I assume the forms cache needs to be updated upon teleporting. The initial character load is fine, but upon entering an inside area it is unable to find the right ID. (I've tested it with a chicken by manually moving my character to the subject above instead of using the COC command.)

image

**Comment by reveriadive** [comment]
Ignore that for now, accidently merged into the wrong branch.

**Comment by reveriadive** [comment]
Closing this as actors still duplicate with mods applied, and fixing the branch is too complicated. (Over-cluttered at the moment.)
I'll rebase from the original, and open another pull request in the future. The changes here up to ef33580 can be applied if that's possible.

**Pull request closed without merging**